src/service/dynamic_report/dynamic_report.py

```diff
@@ -0,0 +1,179 @@
+        data = self.investment_report.repository.get_base_metrics(
+            int(investment_year), **conditions
+        )
+
+        self.investment_report.add_calculated_metrics(data, access)
+        data = self.investment_report.report.filter_by_conditions(data, **conditions)
+
+        return data
+
+    def __get_dynamic_metrics(
+        self,
+        metric: str,
+        calendar_year: int,
+        invest_year: int,
+        access: bool,
+        username,
+        **conditions,
+    ) -> dict:
+        companies = dict()
+        filters = self.metric_report.repository.add_company_filters(**conditions)
+        self.metric_report.company_anonymization.set_company_permissions(username)
+        years = [calendar_year] if calendar_year else []
+
+        if invest_year is not None:
+            investments = self.investment_report.repository.get_investments()
+            companies_with_investments = investments.keys()
+            invest_year_by_company = {
+                company_id: investments.get(company_id)["invest_year"] + invest_year
+                for company_id in companies_with_investments
+            }
+            years = set(invest_year_by_company.values())
+            filters.update(
+                {
+                    "company.id": [
+                        f"'{company}'" for company in companies_with_investments
+                    ]
+                }
+            )
+
+        data = self.metric_report.get_records(metric, years, filters)
+        profiles, sizes = self.metric_report.get_profiles(filters)
+
+        for id in data:
+            company = data[id]
+            if self.metric_report.is_in_range(profiles.get(id), **conditions):
+                self.metric_report.verify_anonimization(
+                    access,
+                    metric,
+                    company,
+                    sizes,
+                    self.metric_report.company_anonymization.companies,
+                )
+                year_metric = (
+                    invest_year_by_company.get(id)
+                    if invest_year is not None
+                    else years[0]
+                )
+                metric_value = {
+                    metric: value
+                    for key, value in company.get("metrics").items()
+                    if key == year_metric
+                }
+
+                company.pop("metrics")
+                company.update(metric_value)
+                companies[id] = company
+
+        return companies
+
+    def get_dynamic_report(
+        self,
+        company_id: str,
+        username: str,
+        metric: str,
+        calendar_year: int,
+        investment_year: int,
+        from_main: bool,
+        access: bool,
+        **conditions,
+    ) -> dict:
+        try:
+            company = dict()
+            is_valid_company = company_id and company_id.strip()
+            header = []
+            data = []
+
+            if self.__is_by_year_report(
```

Comment:
Investment year report and calendar year report could be on another function "get_year_report", and you can verify if it is calendar or investment year, and return that report.
So it will be something like if is_by_year_report: return call function
And inside that function, you can see if you should call investment_report or calendar_report, because it's the same header and same dictionary result structure.